src/Rings/MPolyMap/AffineAlgebras.jl

```diff
@@ -0,0 +1,220 @@
+
+affine_algebra_morphism_type(::Type{T}) where {T <: Union{MPolyRing, MPolyQuo}} = MPolyAnyMap{T, T, Nothing, elem_type(T)}
+
+affine_algebra_morphism_type(R::T) where {T} = affine_algebra_morphism_type(T)
+
+affine_algebra_morphism_type(::Type{S}, ::Type{T}) where {S <: Union{MPolyRing, MPolyQuo}, T <: Union{MPolyRing, MPolyQuo}} = MPolyAnyMap{S, T, Nothing, elem_type(T)}
+
+affine_algebra_morphism_type(R::S, U::T) where {S, T} = affine_algebra_morphism_type(S, T)
+
+################################################################################
+#
+#  Singular data stuff
+#
+################################################################################
+
+function _singular_ring_domain(f::MPolyAnyMap)
+  return get_attribute!(f, :singular_ring_domain) do
+    singular_ring(domain(f))
+  end
+end
+
+function _singular_ring_codomain(f::MPolyAnyMap)
+  return get_attribute!(f, :singular_ring_codomain) do
+    singular_ring(codomain(f))
+  end
+end
+
+function _singular_algebra_morphism(f::MPolyAnyMap)
+  return get_attribute!(f, :singular_algebra_morphism) do
+    DS = _singular_ring_domain(f)
+    CS = _singular_ring_codomain(f)
+    CSimgs = CS.(_images(f))
+    return Singular.AlgebraHomomorphism(DS, CS, CSimgs)
+  end
+end
+
+################################################################################
+#
+#  Preimage of ideal
+#
+################################################################################
+
+function preimage(f::AffAlgHom, I::Union{MPolyIdeal, MPolyQuoIdeal})
+  @req base_ring(I) === codomain(f) "Parent mismatch"
+  D = domain(f)
+  salghom = _singular_algebra_morphism(f)
+  CS = codomain(salghom)
+  V = gens(I)
+  Ix = Singular.Ideal(CS, CS.(V))
+  prIx = Singular.preimage(salghom, Ix)
+  return ideal(D, D.(gens(prIx)))
+end
+
+################################################################################
+#
+#  Kernel
+#
+################################################################################
+
+function kernel(f::AffAlgHom)
+  get_attribute!(f, :kernel) do
+    C = codomain(f)
+    return preimage(f, ideal(C, [zero(C)]))
+  end # TODO: need some ideal_type(domain(f)) here :)
+end
+
+##############################################################################
+#
+#  Injectivity
+#
+##############################################################################
+
+@doc Markdown.doc"""
+    isinjective(F::AlgHom)
+
+Return `true` if `F` is injective, `false` otherwise.
+"""
+function isinjective(F::AffAlgHom)
+   G = gens(kernel(F))
+   for i in 1:length(G)
+      !iszero(G[i]) && return false
+   end
+   return true
+end
+
+# Helper function related to the computation of surjectivity, preimage etc.
+# Stores the necessary data in groebner_data, resp. groebner_data_lex
+function _groebner_data(F::AffAlgHom, ord::Symbol)
+   r = domain(F)
+   s = codomain(F)
+   n = ngens(r)
+   m = ngens(s)
+   return get_attribute!(F, ord) do
```

Comment:
```suggestion
   return get_attribute!(F, ord) do
     r = domain(F)
     s = codomain(F)
     n = ngens(r)
     m = ngens(s)
```
No point in querying these values if you don't need them. Plus this way the code in the do-block does not need to access these "outside" variables.